crates/kubelet/src/node.rs

```diff
@@ -166,6 +168,120 @@ pub async fn create_node<P: 'static + Provider + Sync + Send>(
     info!("Successfully created node '{}'", &config.node_name);
 }
 
+/// Fetch the uid of a node by name.
+pub async fn node_uid(client: &kube::Client, node_name: &str) -> anyhow::Result<String> {
+    let node_client: Api<KubeNode> = Api::all(client.clone());
+    match retry!(node_client.get(node_name).await, times: 4, log_error: |e| error!("Failed to get node to cordon: {:?}", e))
+    {
+        Ok(node) => match node.metadata.and_then(|meta| meta.uid) {
+            Some(uid) => Ok(uid),
+            None => {
+                let err = format!("Node missing metadata or uid {}.", node_name);
+                error!("{}", &err);
+                anyhow::bail!(err);
+            }
+        },
+        Err(e) => {
+            error!("Error fetching node {} id: {:?}.", node_name, e);
+            anyhow::bail!(e);
+        }
+    }
+}
+
+/// Cordons node and evicts all pods.
+pub async fn drain_node(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    cordon_node(client, node_name).await?;
+    evict_pods(client, node_name).await?;
+    Ok(())
+}
+
+/// Fetches list of pods on this node and deletes them.
+pub async fn evict_pods(client: &kube::Client, node_name: &str) -> anyhow::Result<()> {
+    let pod_client: Api<KubePod> = Api::all(client.clone());
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let kube::api::ObjectList { items: pods, .. } = pod_client.list(&params).await?;
+
+    let lp = ListParams::default().fields(&format!("spec.nodeName={}", node_name));
+
+    // The delete call may return a "pending" response, we must watch for the actual delete event.
+    // TODO I dont know what the version number does here.
+    let mut stream = pod_client.watch(&lp, "0").await?.boxed();
+
+    warn!("Evicting {} pods.", pods.len());
+
+    // TODO: Filter mirror pods,daemonsets, kube-system?
```

Comment:
For feature parity with what `kubectl drain` does, it should _technically_ error if it can't remove these kinds of pods (except for mirror, those are ignored). However, we probably want the `--ignore-daemonset` equivalent and don't care if we error here anyway.

tl;dr: Ignore mirror and daemonsets. Kube-system should be evicted if they are not daemonsets

Reply:
Ok, I will give this a shot. In my testing, kube-proxy gets rescheduled onto the node and can be left as a zombie. Do you think it would be reasonable to have `pod_informer` drop add pod events after shutdown has started to prevent this scheduling?

Reply:
If you can find a nice way to let the pod informer know, I think that is a good idea